vikunja-desktop: init at 0.24.6

kolaente opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR adds the [desktop version](https://kolaente.dev/vikunja/desktop) of [Vikunja](https://vikunja.io). Since it is only an electron wrapper around the frontend and the frontend is already packaged in nixos, this package reuses that.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

kolaente commented:
Not sure about my change to the release notes, never done that before.

mweinelt commented:
The package is not referenced by any top-leve attribute.

mweinelt commented:
> {"error":"invalid_request","error_description":"The request is missing a required parameter, includes an invalid parameter value, includes a parameter more than once, or is otherwise malformed. The 'redirect_uri' parameter does not match any of the OAuth 2.0 Client's pre-registered redirect urls."}

The app recognized my instance was using SSO to login, but once I click that button I get this error.

Tried both `https://example.com` as well as `https://example.com/api/v1`.

kolaente commented:
What frontend version is that? You should be able to see that by opening the dev console in the electron shell.

mweinelt commented:
The deployment still uses 0.19.1.

> Vikunja frontend version 0.19.1

kolaente commented:
> The deployment still uses 0.19.1.

Let's wait until the frontend update got merged, that should fix the problem.

> The redirect url is way off.

I'm afraid there's no way around that. The desktop app spins up a local server to serve frontend files, a redirect back from the provider has to happen there.

mweinelt commented:
Please rebase and squash the commits.

kolaente commented:
Took me a little longer than I'd like to admin but it should all be correctly rebased now.

kolaente commented:
@mweinelt I'm not sure what I need to change so that ofborg works.

kolaente committed:
vikunja-desktop: init at 0.24.6

kolaente commented on pkgs/by-name/vi/vikunja-desktop/default.nix:
```diff
@@ -0,0 +1,136 @@
+    hash = "sha256-UT2afhjEangilvflmxxahj7pEiJUWxqUL1Eni1JvuRI=";
+  };
+
+  frontend = stdenv.mkDerivation (finalAttrs: {
```
This is duplicated with the vikunja main package. I wonder if we could resonably abstract this?

mweinelt replied:
Either make it a top-level attribute, or put it into the passthru attrset of another package and consume it as other_package.frontend.

kolaente replied:
Done!

kolaente commented:
Rebased and updated. The package compiles and starts successfully, but then fails with this error when running it:

```
App threw an error during load
Error: Cannot find module 'body-parser'
Require stack:
- /nix/store/y18anvcdqk1qk406v0m93i74gpva6f1l-vikunja-desktop-0.24.3/share/lib/vikunja-desktop/resources/app.asar/node_modules/express/lib/express.js
- /nix/store/y18anvcdqk1qk406v0m93i74gpva6f1l-vikunja-desktop-0.24.3/share/lib/vikunja-desktop/resources/app.asar/node_modules/express/index.js
- /nix/store/y18anvcdqk1qk406v0m93i74gpva6f1l-vikunja-desktop-0.24.3/share/lib/vikunja-desktop/resources/app.asar/main.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1152:15)
    at s._resolveFilename (node:electron/js2c/browser_init:2:121381)
    at Module._load (node:internal/modules/cjs/loader:993:27)
    at c._load (node:electron/js2c/node_init:2:17025)
    at Module.require (node:internal/modules/cjs/loader:1240:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/nix/store/y18anvcdqk1qk406v0m93i74gpva6f1l-vikunja-desktop-0.24.3/share/lib/vikunja-desktop/resources/app.asar/node_modules/express/lib/express.js:15:18)
    at Module._compile (node:internal/modules/cjs/loader:1373:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1432:10)
    at Module.load (node:internal/modules/cjs/loader:1215:32)
```

I have absolutely no idea why it does this. The packages compiled by upstream work fine, so this does not seem to be a problem of dependency declaration here.

After having spent a few hours on this without anything really to show for, I wonder if it would be easier to fetch the built package by upstream and patch it?

mweinelt commented:
Will defer to @lilyinstarlight and @winterqt

maralorn commented:
I would love to see progress on this, but I have little to no clue about the problem.

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+  pname = finalAttrs.name;
+  inherit version src;
+
+  sourceRoot = "${finalAttrs.src.name}/desktop";
```
We need the flag for pnpm install '--shamefully-hoist' since per default pnpm does not install its packages to `node_modules` see [this](https://github.com/electron-userland/electron-builder/issues/6289#issuecomment-1042620422). From my [understanding](https://pnpm.io/cli/install#--shamefully-hoist) it will symlink all packages to `node_modules`.

weriomat replied:
```suggestion
  sourceRoot = "${finalAttrs.src.name}/desktop";
  pnpmInstallFlags = [ "--shamefully-hoist" ];
```

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+      pname
+      version
+      src
+      sourceRoot
```
We also need to pass the Argument to `pnpmDeps` according to [this](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/javascript.section.md#pnpm-javascript-pnpm)

weriomat replied:
```suggestion
      sourceRoot
      pnpmInstallFlags
```

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+
+    sed -i "s/\$${version}/${version}/g" package.json
+    sed -i "s/\"version\": \".*\"/\"version\": \"${version}\"/" package.json
+
```
From my testing I noticed that we need to symlink the frontend before building the asar package. This may be a bad solution to the problem, I have no idea what I am doing ;)
From my understanding and looking at the extracted asar package I noticed that the `frontend` directory is copied (the symlink is resolved), this will increase the binary size by about `18MB`.

weriomat replied:
```suggestion
    ln -s '${vikunja.passthru.frontend}' frontend
```

kolaente replied:
But don't we need the frontend files in the binary in any case?

weriomat replied:
Honestly I have no clue, I have never worked with either JavaScript nor electron.

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+    mkdir -p "$out/share/lib/vikunja-desktop"
+    cp -r ./dist/*-unpacked/{locales,resources{,.pak}} "$out/share/lib/vikunja-desktop"
+    cp -r ./node_modules "$out/share/lib/vikunja-desktop/resources"
+    ln -s '${vikunja.passthru.frontend}' "$out/share/lib/vikunja-desktop/resources/frontend"
```
Since we symlinked earlier we can remove this symlink

weriomat replied:
```suggestion
```

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+  stdenv,
+  makeWrapper,
+  makeDesktopItem,
+  callPackage,
```
This is unused, so we can remove this.

weriomat replied:
```suggestion
```

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+  electron,
+  unstableGitUpdater,
+  fetchFromGitHub,
+  fetchzip,
```
This is unused as well

weriomat replied:
```suggestion
```

weriomat reviewed (changes requested):
I got it to build locally, but I am not satisfied with the solution of symlinking the frontend before building the `app.asar`. There might be a better solution, but I don't have enough knowledge to provide one. 
I would love to see progress on this and get this merged :)

On a side note, the current (25.05) release notes should probably be updated instead of the 24.11 one.

gepbird commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,119 @@
+  makeWrapper,
+  makeDesktopItem,
+  callPackage,
+  pnpm,
```
Most likely you need to change `pnpm` to `pnpm_9`. Since #371832, it points to `pnpm_10` instead of `pnpm_9`.

kolaente replied:
Done.

weriomat commented:
On another note, since we do have a monorepo it might be an okay to `passthru` the `src` and `version` in order to not have to bump the version and hash of both `vikunja` and `vikunja-desktop`. Any opinions on this idea?

kolaente commented:
Thanks for your suggestions!

> On another note, since we do have a monorepo it might be an okay to `passthru` the `src` and `version` in order to not have to bump the version and hash of both `vikunja` and `vikunja-desktop`. Any opinions on this idea?

I like that idea, also because then we can update the desktop and server packages in one go.

kolaente commented:
I've applied the changes and upgraded to 0.24.6. It seems to build and work fine now. 

Can I get another review from @lilyinstarlight, @winterqt and maybe @mweinelt or @leona-ya (since you're one of the maintainers of the Vikunja server package)?

mweinelt commented:
Starts like that:

![image](https://github.com/user-attachments/assets/d415f0b1-0a12-4d8a-a10b-1ad6c73728bf)

Entering my instance (https://todo.darmstadt.ccc.de/api/v1) and clicking `CHANGE`. No visual feedback.

For each click on `CHANGE` I see a successful HTTP request against the API, but it will not proceed.

```
"GET /api/v1/info HTTP/2.0" 200 424 "http://127.0.0.1:45735/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) vikunja-desktop/0.24.6 Chrome/130.0.6723.170 Electron/33.3.1 Safari/537.36"

kolaente commented:
@mweinelt Looks like you didn't [enable cors](https://vikunja.io/docs/config-options/#0--cors) for your instance.

If you only want to allow connections from Vikunja desktop, you need to enable at least `http://localhost:45735` (if that port is taken, it will try `port++` until it finds one that's free). Since you're using openid auth, you'll need to make sure that the provider allows redirects back to that host as well.

I was able to verify that the desktop app works with [the demo](https://try.vikunja.io).

mweinelt commented:
Why does Vikunja itself not send CORS headers when accessing the API though?

kolaente commented:
> Why does Vikunja itself not send CORS headers when accessing the API though?

Because the API can't know in which context it is accessed. That's why each host needs to be allowed explicitly.

A workaround could be to allow access from localhost by default, but that's a Vikunja issue.

kolaente commented:
> Can confirm accesing it works after allowing CORS, but now I'm stuck at logging into the OIDC provider using my Passkey 🤡

That sounds like a Vikunja issue as well, not sure if that belongs here?

mweinelt commented:
Yeah, sorry for mixing issues with our setup and reviewing/testing the app.

weriomat commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```diff
@@ -0,0 +1,118 @@
+    description = "Desktop App of the Vikunja to-do list app";
+    homepage = "https://vikunja.io/";
+    license = licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [ kolaente ];
```
Maybe add the `mainProgram` here as well

weriomat replied:
```suggestion
    maintainers = with lib.maintainers; [ kolaente ];
    mainProgram = "vikunja-desktop";
```

mweinelt commented on pkgs/by-name/vi/vikunja-desktop/package.nix:
```patch
diff --git a/pkgs/by-name/vi/vikunja-desktop/package.nix b/pkgs/by-name/vi/vikunja-desktop/package.nix
index 00207debc0e2..0e478625d529 100644
--- a/pkgs/by-name/vi/vikunja-desktop/package.nix
+++ b/pkgs/by-name/vi/vikunja-desktop/package.nix
@@ -4,10 +4,12 @@
   makeWrapper,
   makeDesktopItem,
   pnpm,
+  pnpmConfigHook,
   nodejs,
   electron,
   unstableGitUpdater,
   fetchFromGitHub,
+  fetchPnpmDeps,
   vikunja,
 }:
 
@@ -29,7 +31,7 @@ stdenv.mkDerivation (finalAttrs: {
   sourceRoot = "${finalAttrs.src.name}/desktop";
   pnpmInstallFlags = [ "--shamefully-hoist" ];
 
-  pnpmDeps = pnpm.fetchDeps {
+  pnpmDeps = fetchPnpmDeps {
     inherit (finalAttrs)
       pname
       version
@@ -48,7 +50,8 @@ stdenv.mkDerivation (finalAttrs: {
   nativeBuildInputs = [
     makeWrapper
     nodejs
-    pnpm.configHook
+    pnpm
+    pnpmConfigHook
     vikunja.passthru.frontend
   ];
 ```

Works for me after a rebase.

gepbird replied:
Interesting, it starts building for me. Have you pulled latest master from NixOS/nixpkgs, so not your fork?

This is what I did:
```console
nixpkgs ❯ git remote -v
d	git@github.com:gepbird/nixpkgs (fetch)
d	git@github.com:gepbird/nixpkgs (push)
origin	git@github.com:NixOS/nixpkgs (fetch)
origin	git@github.com:NixOS/nixpkgs (push)

nixpkgs ❯ gh pr checkout 195231
From github.com:NixOS/nixpkgs
 * [new ref]                   refs/pull/195231/head -> feature/add-vikunja-desktop

nixpkgs ❯ git fetch origin
From github.com:NixOS/nixpkgs
 * branch                      refs/pull/195231/head -> FETCH_HEAD

nixpkgs ❯ git rebase origin/master
Successfully rebased and updated refs/heads/feature/add-vikunja-desktop.
Switched to branch 'feature/add-vikunja-desktop'

nixpkgs ❯ # make the changes

nixpkgs ❯ git diff
diff --git a/pkgs/by-name/vi/vikunja-desktop/package.nix b/pkgs/by-name/vi/vikunja-desktop/package.nix
index 00207debc0e2..120528a97b99 100644
--- a/pkgs/by-name/vi/vikunja-desktop/package.nix
+++ b/pkgs/by-name/vi/vikunja-desktop/package.nix
@@ -9,6 +9,7 @@
   unstableGitUpdater,
   fetchFromGitHub,
   vikunja,
+  fetchPnpmDeps,
 }:

 let
@@ -29,7 +30,7 @@ stdenv.mkDerivation (finalAttrs: {
   sourceRoot = "${finalAttrs.src.name}/desktop";
   pnpmInstallFlags = [ "--shamefully-hoist" ];

-  pnpmDeps = pnpm.fetchDeps {
+  pnpmDeps = fetchPnpmDeps {
     inherit (finalAttrs)
       pname
       version

nixpkgs ❯ nix build -f . vikunja-desktop
trace: evaluation warning: pnpm.configHook: The package attribute is deprecated. Use the top-level pnpmConfigHook attribute instead

nixpkgs ❯
```
